packages/flutter_tools/lib/src/ios/xcode_debug.dart

```diff
@@ -0,0 +1,498 @@
+  }
+
+  String get pathToXcodeAutomationScript {
+    final String flutterToolsAbsolutePath = _fileSystem.path.join(
+      _flutterRoot,
+      'packages',
+      'flutter_tools',
+    );
+    return '$flutterToolsAbsolutePath/bin/xcode_debug.js';
+  }
+
+  /// Install, launch, and start a debug session for app through Xcode interface,
+  /// automated by OSA scripting. First checks if the project is opened in
+  /// Xcode. If it isn't, open it with the `open` command.
+  ///
+  /// The OSA script waits until the project is opened and the debug action
+  /// has started. It does not wait for the app to install, launch, or start
+  /// the debug session.
+  Future<bool> debugApp({
+    required XcodeDebugProject project,
+    required String deviceId,
+    required List<String> launchArguments,
+  }) async {
+
+    // If project is not already opened in Xcode, open it.
+    if (!await _isProjectOpenInXcode(project: project)) {
+      final bool openResult = await _openProjectInXcode(xcodeWorkspace: project.xcodeWorkspace);
+      if (!openResult) {
+        return openResult;
+      }
+    }
+
+    currentDebuggingProject = project;
+    StreamSubscription<String>? stdoutSubscription;
+    StreamSubscription<String>? stderrSubscription;
+    try {
+      startDebugActionProcess = await _processUtils.start(
+        <String>[
+          ..._xcode.xcrunCommand(),
+          'osascript',
+          '-l',
+          'JavaScript',
+          pathToXcodeAutomationScript,
+          'debug',
+          '--xcode-path',
+          pathToXcodeApp,
+          '--project-path',
+          project.xcodeProject.path,
+          '--workspace-path',
+          project.xcodeWorkspace.path,
+          '--device-id',
+          deviceId,
+          '--scheme',
+          project.scheme,
+          '--skip-building',
+          '--launch-args',
+          json.encode(launchArguments),
+          if (project.verboseLogging) '--verbose',
+        ],
+      );
+
+      String stdout = '';
+      stdoutSubscription = startDebugActionProcess!.stdout
+          .transform<String>(utf8.decoder)
+          .transform<String>(const LineSplitter())
+          .listen((String line) {
+            _logger.printTrace(line);
+            stdout += line;
+      });
+
+      String stderr = '';
+      // console.log from the script are found in the stderr
+      stderrSubscription = startDebugActionProcess!.stderr
+          .transform<String>(utf8.decoder)
+          .transform<String>(const LineSplitter())
+          .listen((String line) {
+            _logger.printTrace('stderr: $line');
+            stderr += line;
+      });
+
+      final int exitCode = await startDebugActionProcess!.exitCode.whenComplete(() async {
+        await stdoutSubscription?.cancel();
+        await stderrSubscription?.cancel();
+        startDebugActionProcess = null;
+      });
+
+      if (exitCode != 0) {
+        _logger.printError('Error executing osascript: $exitCode\n$stderr');
```

Comment:
does osascript log anything interesting to stdout? I haven't tested osascript, but sometimes for CLI tools stdout will have helpful context as to what it was trying to do when the error occurred.

Reply:
From my debugging, it doesn't appear to send anything to stdout when it errors

Reply:
:+1: